### PR TITLE
Blackhole Loud Box Bringup

### PR DESCRIPTION
**Pull request by ign-ssunil**
This PR contains fixes for the below models to run on Blackhole Loudbox. The current state of the models on Blackhole Loudbox.
| Sl.No | Model                                             | Demo | Unit Tests |
|------:|--------------------------------------------------|------|-----------|
| 1 | Mistral 3.1 24B | ✅ | ✅ |
| 2 | Qwen2.5 VL 72B | ✅ | ✅ |
| 3 | GPT-OSS 20B | ✅ | ⚠️ |
| 4 | GPT-OSS 120B | ✅ | ⚠️ |
| 5 | Mochi | ✅ | ⚠️ |
| 6 | Qwen3 Embedding (8B / 6B / 4B) | ✅ | ➖ |
| 7 | Llama3 8B (TT Transformer) | ✅ | ✅ |
| 8 | DeepSeek V3 685B | ❌ | ❌ |
| 9 | Qwen2.5 VL 32B Instruct (Non-TT Transformer) | ✅ | ✅ |
| 10 | Qwen2.5 72B Instruct (TT Transformer) | ✅ | ✅ |
| 11 | Qwen3 VL 32B | ✅ | ✅ |
| 12 | Qwen2.5 32B Instruct (TT Transformer) | ✅ | ✅ |
| 13 | Qwen3 32B (TT Transformer) | ✅ | ✅ |

 ⚠️  -> Unit test failing due to not meeting mesh topology requirements.
  ❌   -> Supports only Galaxy (WH) - 2x or 4x configurations
➖ -> Unit tests not yet written
Notes:
1. **GPT OSS(20B/120B)**
  - The test models/demos/gpt_oss/tests/fused_op_unit_tests/test_gpt_oss_all_fused_ops_device_perf.py fails in BHLB, The test is supported and run on galaxy in [CI](https://github.com/tenstorrent/tt-metal/actions/runs/22996945420/job/66772421141#step:9:254)
 2. **Mochi**
- pytest models/tt_dit/tests/models/mochi/test_vae_mochi.py -k "4links" fails with requested link index 2 is out of bounds in BHLB
- With the latest build for a few tests got error related group_norm .Here's the link to the same error observed in [CI:](https://github.com/tenstorrent/tt-metal/actions/runs/23029854638/job/66886283419#step:7:380)
3. **Qwen3 embedding**
  - ran using python tt-inference-server/tt-vllm-plugin/examples/qwen3_embedding_8b.py 
  - `export HF_MODEL=Qwen/Qwen3-Embedding-4B` (use 0.6B or 8B) Ensure the same model is being used in tt-inference-server/tt-vllm-plugin/examples/qwen3_embedding_8b.py and tt-metal/models/demos/wormhole/qwen3_embedding_8b/generator_vllm.py

- The branch has been rebased and models have been verified in [commit](https://github.com/tenstorrent/tt-metal/commit/9708a8edae14b9327ed61b610972716733311b7c) 
### Checklist

- [ ] [Sanity tests](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:ign/qwen3_32b)
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:ign/qwen3_32b)
- [ ] [cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:ign/qwen3_32b)
- [ ] New/Existing tests provide coverage for changes
- [ ] (Optional) Ran [clang-tidy code analysis](https://github.com/tenstorrent/tt-metal/actions/workflows/code-analysis.yaml) on the PR branch to catch linting errors early

#### Model tests

If your changes cover model-related code, you should run tests corresponding to affected models and platforms (Single card, T3K, Galaxy). "Choose your pipeline" workflows facilitate running multiple kinds of tests in a single run. Each offers `models-mandatory` and `models-extended` presets.
The former includes a minimal set of tests, to be run always. The latter extends that with additional ones - use your best judgement in deciding which is the most appropriate for your PR.

- [ ] [(Single) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:ign/qwen3_32b)
  - [ ] `models-mandatory` preset (runs: [Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) and [Frequent model and ttnn tests](https://github.com/tenstorrent/tt-metal/actions/workflows/fast-dispatch-full-regressions-and-models.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [(T3K) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:ign/qwen3_32b)
  - [ ] `models-mandatory` preset (runs: [Unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-model-perf-tests.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [(Galaxy) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:ign/qwen3_32b)
  - [ ] `models-mandatory` preset (runs: [Quick tests](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-perf-tests.yaml) tests)
  - [ ] other selection - specify runs  - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]   - [ ]

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:ign/bh-lb-bringup)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:ign/bh-lb-bringup)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:ign/bh-lb-bringup)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:ign/bh-lb-bringup)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:ign/bh-lb-bringup)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:ign/bh-lb-bringup)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:ign/bh-lb-bringup)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:ign/bh-lb-bringup)

### CODE REVIEW

File: models/tt_transformers/tt/lm_head.py [hunk 1]
```diff
@@ -35,6 +35,14 @@ def __init__(
...
         self.prefetcher = prefetcher
 
+        # Pad vocab_size to be divisible by (32 * num_devices) so that:
+        # 1. vocab_size is tile-aligned (divisible by 32)
+        # 2. size_per_device is also tile-aligned after dividing by num_devices
+        # This ensures TILE concat doesn't have padding in the middle
+        tile_size = 32
+        self.padded_vocab_size = math.ceil(self.padded_vocab_size / (tile_size * self.num_devices)) * (
+            tile_size * self.num_devices
+        )
```
Comment (gwangTT) [inline models/tt_transformers/tt/lm_head.py]: On line 34, self.padded_vocab_size is receiving value from args.padded_vocab_size. I think this padding logic better belongs in `model_config.py`. model_config.py should be already doing this

**Review by gwangTT — COMMENTED** [review]
Submitting one comment. Still want to look at changes in `generator.py` more closely. Will do that soon. The rest looks good to me.

**Comment by zzigler-tt** [comment]
@feslamiTT Can you please work with the appropriate folks to push this PR through? Thanks

File: models/tt_transformers/tt/lm_head.py (hunk 1 above)
Comment (ign-saurav) [reply models/tt_transformers/tt/lm_head.py]: @gwangTT , Removed the padding logic from lm_head.py.

File: models/tt_transformers/tt/generator.py [hunk 2]
```diff
@@ -723,15 +723,36 @@ def prefill_forward_text(
...
+                            slot_hidden_list.append((slot, slot_hidden, last_token_idx[slot]))
+                        ttnn.synchronize_device(self.model[model_id].mesh_device)
+                        dim = self.model[model_id].args.dim
+                        for slot, slot_hidden, lt_idx in slot_hidden_list:
+                            slot_hidden_torch = ttnn.to_torch(ttnn.get_device_tensors(slot_hidden)[0]).float()
+                            pos = int(lt_idx % 32)
+                            out = slot_hidden_torch[0, 0, pos, :dim].clone()
+                            if out.device.type != "cpu":
+                                out = out.cpu()
+                            output_tensor[slot] = out
```
Comment (gwangTT) [inline models/tt_transformers/tt/generator.py:764]: Can you please explain what this added code is used for?

Comment (ign-saurav) [reply models/tt_transformers/tt/generator.py:764]: This code adds support for Qwen3 embedding models (0.6B/4B/8B), which output a hidden state vector instead of vocabulary logits. In `generator.py`, the batched prefill trace path is used to produce vocabulary logits. With this change, the code now returns the last token’s hidden state after the final layer normalization as a [hidden_dim] embedding vector, skipping the LM head.

**Comment by apande-TT** [comment]
I am attaching the link to regression run https://github.com/tenstorrent/tt-metal/actions/runs/24349581242. I have checked the failures against the CI runs on the  latest main. Looks good to me

**Review by gwangTT — APPROVED** [review]
LGTM. Thanks for adding support for BH LB!

**Review by jyuanTT — APPROVED** [review]

**Comment by jyuanTT** [comment]
Is it still a draft?

**Review by uaydonat — APPROVED** [review]

**Comment by mtairum** [comment]
@ign-ssunil please rebase to latest main and run the new models CI https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml. 

Right now this only runs the WH variants, so does not substitute the new BH tests you've added, but since you are making changes to the TTT generator, we need to make sure that we do not observe regressions. 

on that pipeline, run Tier 1 + 2 + 3,  e2e + unit tests.

**Comment by ign-saurav** [comment]
> @ign-ssunil please rebase to latest main and run the new models CI https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml.
> 
> Right now this only runs the WH variants, so does not substitute the new BH tests you've added, but since you are making changes to the TTT generator, we need to make sure that we do not observe regressions.
> 
> on that pipeline, run Tier 1 + 2 + 3, e2e + unit tests.

Tier1+2+3, e2e + unit test CI run - https://github.com/tenstorrent/tt-metal/actions/runs/24775770391

**Comment by mtairum** [comment]
@ign-saurav Please note that I just gave a quick look at the status of the unit pipelines on schedule CI and they seem to be in a dire straight. Not sure yet what broke  them. They were ok 2 days ago 😭 


For reference:
- A scheduled tier 2 unit: https://github.com/tenstorrent/tt-metal/actions/runs/24811954783
- A scheduled tier 3 unit: https://github.com/tenstorrent/tt-metal/actions/runs/24764361448

**Comment by mtairum** [comment]
The issue was a recent UMD bump. I'll ping this thread when the fix/revert hits main and you can rebase and re-test then.

**Comment by mtairum** [comment]
No issues in main (that I know off). Feel free to rebase and retest the Tier1+2+3, e2e + unit test CI run 🙏

**Comment by ign-saurav** [comment]
> No issues in main (that I know off). Feel free to rebase and retest the Tier1+2+3, e2e + unit test CI run 🙏

CI run - https://github.com/tenstorrent/tt-metal/actions/runs/24883968758

**Comment by mtairum** [comment]
When the CI run finishes, you can compare to [this one ](https://github.com/tenstorrent/tt-metal/actions/runs/24889404538)I ran today off main. 
If there's no added issues on top, all's good and I'll approve 👍

**Review by mtairum — APPROVED** [review]

**Comment by mtairum** [comment]
@ign-ssunil anything missing still, or can we merge?

**Comment by ign-saurav** [comment]
> @ign-ssunil anything missing still, or can we merge?

@mtairum , We can merge the PR. 
The CI runs are showing similar results that of main and the failures noticed are not related to the PR changes.

**Comment by mtairum** [comment]
Added to merge queue. Please keep an eye on Blackhole pipelines for the next days in case something regresses.
